### PR TITLE
Add discard feature

### PR DESCRIPTION
**Pull request by RexJaeschke**
This PR addresses the discard part of Issue #561 as well as the use of a discard on the lhs of simple assignment, which was previously covered by PR #44. That text will be removed from PR 44, and that PR will be revised to reflect PR 596's existence.

I have yet to review PR #61 or #63 in terms of this new PR. When I eventually do so, I might need to come back and tweak this PR.

The local-variable creation part of Issue #561 will be addressed separately.

### CODE REVIEW

File: standard/expressions.md [hunk 1]
```diff
@@ -5678,7 +5678,7 @@ assignment_operator
 
 The left operand of an assignment shall be an expression classified as a variable, a property access, an indexer access, or an event access.
 
-The `=` operator is called the ***simple assignment operator***. It assigns the value of the right operand to the variable, property, or indexer element given by the left operand. The left operand of the simple assignment operator shall not be an event access (except as described in [§14.8.2](classes.md#1482-field-like-events)). The simple assignment operator is described in [§11.18.2](expressions.md#11182-simple-assignment).
+The `=` operator is called the ***simple assignment operator***. When the left operand is not a discard (§discards-new-clause), this operator assigns the value of the right operand to the variable, property, or indexer element given by the left operand. Otherwise, the right operand is evaluated with its value being discarded. The left operand of the simple assignment operator shall not be an event access (except as described in [§14.8.2](classes.md#1482-field-like-events)). The simple assignment operator is described in [§11.18.2](expressions.md#11182-simple-assignment).
```
Comment (Nigel-Ecma) [inline standard/expressions.md]: Should we make it more obvious that the right operand is always evaluated by stating that first? Something along the lines of:

> The `=` operator is called the ***simple assignment operator***. This evaluates the right operand and then either: assigns the value to the variable, property, or indexer element given by the left operand; or discards the value if the left operand is a discard (§discards-new-clause). The left operand of the simple assignment operator shall not be an event access (except as described in [§14.8.2](classes.md#1482-field-like-events)). The simple assignment operator is described in [§11.18.2](expressions.md#11182-simple-assignment).

File: standard/variables.md [hunk 2]
```diff
@@ -817,3 +817,21 @@ variable_reference
...
+
+A discard is represented by the identifier `_`.
+
+When `_` represents a discard (rather than a named program element), it is a compile-time error to attempt to retrieve its value.
+
+A discard is not a variable reference ([§9.5](variables.md#95-variable-references)).
+
+Under certain circumstances, a discard may be used in the same scope as an explicitly declared variable named `_`.
+
+For details of using discards when deconstructing tuples, see §deconstruction-expressions-new-clause. When calling methods with `out` parameters, see [§14.6.2.4](classes.md#14624-output-parameters). In a pattern-matching operation with the `is` and `switch` statements, see [§11.11.11](expressions.md#111111-the-is-operator) and [§12.8.3](statements.md#1283-the-switch-statement), respectively. As the left operand of a simple assignment, see [§11.18.2](expressions.md#11182-simple-assignment).
```
Comment (Nigel-Ecma) [inline standard/variables.md]: Maybe a list, textual or bullets? This is just changing punctuation but might be clearer, e.g.

> For details of using discards: when deconstructing tuples, see §deconstruction-expressions-new-clause; when calling methods with `out` parameters, see [§14.6.2.4](classes.md#14624-output-parameters); in a pattern-matching operation with the `is` and `switch` statements, see [§11.11.11](expressions.md#111111-the-is-operator) and [§12.8.3](statements.md#1283-the-switch-statement), respectively; and as the left operand of a simple assignment, see [§11.18.2](expressions.md#11182-simple-assignment).

or

> For details of using discards:
> - when deconstructing tuples, see §deconstruction-expressions-new-clause;
> - when calling methods with `out` parameters, see [§14.6.2.4](classes.md#14624-output-parameters);
> - in a pattern-matching operation with the `is` and `switch` statements, see [§11.11.11](expressions.md#111111-the-is-operator) and [§12.8.3](statements.md#1283-the-switch-statement), respectively; and
> - as the left operand of a simple assignment, see [§11.18.2](expressions.md#11182-simple-assignment).

**Review by Nigel-Ecma — APPROVED** [review]
This looks fine, my two comments are just optional suggestions on presentation not content.

File: standard/variables.md (hunk 2 above)
Comment (BillWagner) [inline standard/variables.md]: This is causing the build issue. The `deconstruction-expression-new-clause` isn't resolving.

**Comment by BillWagner** [comment]
Closed in favor of #664

**Pull request closed without merging**